opnode/node/node.go

```diff
@@ -2,85 +2,175 @@ package node
 
 	// TODO: multi-addrs option
 	// TODO: bootnodes option
 
-	log Logger
+	log log.Logger
+
+	// sources to fetch data from
+	l1Sources []l1.Source
 
-	l1Node   *rpc.Client
-	l2Engine *rpc.Client
+	// engines to keep synced
+	l2Engines []*l2.Engine
+
+	l1Tracker    l1.Tracker
+	l1Downloader l1.Downloader
 
 	ctx   context.Context
 	close chan chan error
 }
 
 func (c *OpNodeCmd) Default() {
-	c.L1NodeAddr = "http://127.0.0.1:8545"
-	c.L2EngineAddr = "http://127.0.0.1:8551"
+	c.L1NodeAddrs = []string{"http://127.0.0.1:8545"}
+	c.L2EngineAddrs = []string{"http://127.0.0.1:8551"}
 }
 
 func (c *OpNodeCmd) Help() string {
 	return "Run optimism node"
 }
 
 func (c *OpNodeCmd) Run(ctx context.Context, args ...string) error {
-	log := c.LogCmd.Create()
-	c.log = log
+	logger := c.LogCmd.Create()
+	c.log = logger
 	c.ctx = ctx
 
-	// L1 exec engine: read-only, to update L2 consensus with
-	l1Node, err := rpc.DialContext(ctx, c.L1NodeAddr)
-	if err != nil {
-		return err
+	for i, addr := range c.L1NodeAddrs {
+		// L1 exec engine: read-only, to update L2 consensus with
+		l1Node, err := rpc.DialContext(ctx, addr)
+		if err != nil {
+			// HTTP or WS RPC may create a disconnected client, RPC over IPC may fail directly
+			if l1Node == nil {
+				return fmt.Errorf("failed to dial L1 address %d (%s): %v", i, addr, err)
+			}
+			c.log.Warn("failed to dial L1 address, but may connect later", "i", i, "addr", addr, "err", err)
+		}
+		// TODO: we may need to authenticate the connection with L1
+		// l1Node.SetHeader()
+		cl := ethclient.NewClient(l1Node)
+		c.l1Sources = append(c.l1Sources, cl)
+	}
+	if len(c.l1Sources) == 0 {
+		return fmt.Errorf("need at least one L1 source endpoint, see --l1")
 	}
-	c.l1Node = l1Node
 
-	// L2 exec engine: updated by this OpNode (L2 consensus layer node)
-	engine, err := rpc.DialContext(ctx, c.L2EngineAddr)
-	if err != nil {
-		return err
+	for i, addr := range c.L2EngineAddrs {
+		// L2 exec engine: updated by this OpNode (L2 consensus layer node)
+		backend, err := rpc.DialContext(ctx, addr)
+		if err != nil {
+			if backend == nil {
+				return fmt.Errorf("failed to dial L2 address %d (%s): %v", i, addr, err)
+			}
+			c.log.Warn("failed to dial L2 address, but may connect later", "i", i, "addr", addr, "err", err)
+		}
+		// TODO: we may need to authenticate the connection with L2
+		// backend.SetHeader()
+		client := &l2.EngineClient{
+			RPCBackend: backend,
+			Log:        c.log.New("engine_client", i),
+		}
+		engine := &l2.Engine{
+			Ctx: c.ctx,
+			Log: c.log.New("engine", i),
+			RPC: client,
+		}
+		c.l2Engines = append(c.l2Engines, engine)
 	}
-	c.l2Engine = engine
 
 	// TODO: maybe spin up an API server
 	//  (to get debug data, change runtime settings like logging, serve pprof, get peering info, node health, etc.)
 
+	c.l1Tracker = l1.NewTracker()
+
 	c.close = make(chan chan error)
 
 	go c.RunNode()
 
 	return nil
 }
 
-// background process
 func (c *OpNodeCmd) RunNode() {
```

Comment:
- What is the distinction between `Run`  and `RunNode`?
- Must `RunNode` be public?